apiserver: `PROJECT_NAME_VERSION_IDX` constraint violations while forwarding data from V4

### Current Behavior

To load-test the newly-migrated V5 test instance under realistic circumstances I set up BOM forwarding via Webhooks as outlined [here](https://dependencytrack.github.io/docs/next/guides/administration/running-v4-and-v5-in-parallel/).

After having run this setup overnight, I found several instances of the following entries in the forwarder (bento) logs:
```
level=error msg="Failed to send message to http_client: http://apiserver-service:8080/api/v1/bom: HTTP request returned unexpected response code (400): 400 Bad Request, Error: [{\"message\":\"must not be blank\",\"messageTemplate\":\"{jakarta.validation.constraints.NotBlank.message}\",\"path\":\"projectVersion\",\"invalidValue\":null}]" @service=bento label="" path=root.output
level=error msg="Failed to send message to http_client: http://apiserver-service:8080/api/v1/bom: HTTP request returned unexpected response code (500): 500 Server Error, Error: Uncaught internal server error" @service=bento label="" path=root.output
level=error msg="Failed to send message to http_client: http://apiserver-service:8080/api/v1/bom: HTTP request returned unexpected response code (400): 400 Bad Request, Error: {\"status\":400,\"title\":\"The provided JSON payload could not be mapped\",\"detail\":\"The BOM is too large to be transmitted safely via Base64 encoded JSON value. Please use the \\\"POST /api/v1/bom\\\" endpoint with Content-Type \\\"multipart/form-data\\\" instead. Original cause: String value length (20051112) exceeds the maximum allowed (20000000, from `StreamReadConstraints.getMaxStringLength()`) (through reference chain: org.dependencytrack.resources.v1.vo.BomSubmitRequest[\\\"bom\\\"])\"}" @service=bento label="" path=root.output
```
Two of these are somewhat expected and related to "bad payloads".
The `Uncaught internal server error` gave me pause, though.

Looking into the apiserver logs resulted in finding something like this for these instances:
```
2026-06-11 07:07:21,411 ERROR [Persist] Insert of object "org.dependencytrack.model.Project@2effebb8" using statement "INSERT INTO "PROJECT" ("AUTHORS","CLASSIFIER","COLLECTION_LOGIC","COLLECTION_TAG_ID","CPE","DESCRIPTION","DIRECT_DEPENDENCIES","EXTERNAL_REFERENCES","GROUP","INACTIVE_SINCE","IS_LATEST","LAST_BOM_IMPORTED","LAST_BOM_IMPORTED_FORMAT","LAST_RISKSCORE","LAST_VULNERABILITY_ANALYSIS","MANUFACTURER","NAME","PARENT_PROJECT_ID","PUBLISHER","PURL","SUPPLIER","SWIDTAGID","UUID","VERSION") VALUES (?,?,?,?,?,?,CAST(? AS JSONB),?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)" failed : ERROR: duplicate key value violates unique constraint "PROJECT_NAME_VERSION_IDX"
  Detail: Key ("NAME", "VERSION")=(name, main) already exists. [principal=odt_zPvz18ZO********************************, requestUri=/v1/bom, requestId=7c021131-d718-44f0-bc06-611d54942dfe, requestMethod=PUT]
2026-06-11 07:07:21,411 ERROR [GlobalExceptionHandler] Uncaught internal server error [principal=odt_zPvz18ZO********************************, requestUri=/v1/bom, requestId=7c021131-d718-44f0-bc06-611d54942dfe, requestMethod=PUT]
javax.jdo.JDODataStoreException: Insert of object "org.dependencytrack.model.Project@2effebb8" using statement "INSERT INTO "PROJECT" ("AUTHORS","CLASSIFIER","COLLECTION_LOGIC","COLLECTION_TAG_ID","CPE","DESCRIPTION","DIRECT_DEPENDENCIES","EXTERNAL_REFERENCES","GROUP","INACTIVE_SINCE","IS_LATEST","LAST_BOM_IMPORTED","LAST_BOM_IMPORTED_FORMAT","LAST_RISKSCORE","LAST_VULNERABILITY_ANALYSIS","MANUFACTURER","NAME","PARENT_PROJECT_ID","PUBLISHER","PURL","SUPPLIER","SWIDTAGID","UUID","VERSION") VALUES (?,?,?,?,?,?,CAST(? AS JSONB),?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)" failed : ERROR: duplicate key value violates unique constraint "PROJECT_NAME_VERSION_IDX"
  Detail: Key ("NAME", "VERSION")=(name, main) already exists.
NestedThrowables:
org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "PROJECT_NAME_VERSION_IDX"
  Detail: Key ("NAME", "VERSION")=(name, main) already exists.
	at org.datanucleus.api.jdo.JDOAdapter.getJDOExceptionForNucleusException(JDOAdapter.java:608)
	at org.datanucleus.api.jdo.JDOPersistenceManager.jdoMakePersistent(JDOPersistenceManager.java:702)
	at org.datanucleus.api.jdo.JDOPersistenceManager.makePersistent(JDOPersistenceManager.java:722)
	at alpine.persistence.AbstractAlpineQueryManager.lambda$persist$0(AbstractAlpineQueryManager.java:284)
	at alpine.persistence.Transaction.call(Transaction.java:139)
	at alpine.persistence.AbstractAlpineQueryManager.callInTransaction(AbstractAlpineQueryManager.java:484)
	at alpine.persistence.AbstractAlpineQueryManager.callInTransaction(AbstractAlpineQueryManager.java:495)
	at alpine.persistence.AbstractAlpineQueryManager.persist(AbstractAlpineQueryManager.java:284)
	at org.dependencytrack.persistence.ProjectQueryManager.lambda$createProject$0(ProjectQueryManager.java:236)
	at alpine.persistence.Transaction.call(Transaction.java:139)
	at alpine.persistence.AbstractAlpineQueryManager.callInTransaction(AbstractAlpineQueryManager.java:484)
	at alpine.persistence.AbstractAlpineQueryManager.callInTransaction(AbstractAlpineQueryManager.java:495)
	at org.dependencytrack.persistence.ProjectQueryManager.createProject(ProjectQueryManager.java:205)
	at org.dependencytrack.persistence.ProjectQueryManager.createProject(ProjectQueryManager.java:192)
	at org.dependencytrack.persistence.QueryManager.createProject(QueryManager.java:450)
	at org.dependencytrack.resources.v1.BomResource.lambda$uploadBom$1(BomResource.java:433)
	at alpine.persistence.Transaction.call(Transaction.java:139)
	at alpine.persistence.AbstractAlpineQueryManager.callInTransaction(AbstractAlpineQueryManager.java:484)
	at alpine.persistence.AbstractAlpineQueryManager.callInTransaction(AbstractAlpineQueryManager.java:495)
	at org.dependencytrack.resources.v1.BomResource.uploadBom(BomResource.java:399)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:146)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:189)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:93)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:470)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:394)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:274)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:266)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:253)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:703)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:422)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:374)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:355)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:309)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:202)
	at org.eclipse.jetty.ee11.servlet.ServletHolder.handle(ServletHolder.java:754)
	at org.eclipse.jetty.ee11.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1668)
	at alpine.server.filters.WhitelistUrlFilter.doFilter(WhitelistUrlFilter.java:165)
	at org.eclipse.jetty.ee11.servlet.FilterHolder.doFilter(FilterHolder.java:205)
	at org.eclipse.jetty.ee11.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1640)
	at org.eclipse.jetty.ee11.servlet.ServletHandler$MappedServlet.handle(ServletHandler.java:1602)
	at org.eclipse.jetty.ee11.servlet.ServletChannel.dispatch(ServletChannel.java:871)
	at org.eclipse.jetty.ee11.servlet.ServletChannel.handle(ServletChannel.java:449)
	at org.eclipse.jetty.ee11.servlet.ServletHandler.handle(ServletHandler.java:469)
	at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1250)
	at org.eclipse.jetty.compression.server.CompressionHandler.handle(CompressionHandler.java:343)
	at org.eclipse.jetty.server.Server.handle(Server.java:197)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:793)
	at org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:419)
	at org.eclipse.jetty.server.internal.HttpConnection$FillableCallback.succeeded(HttpConnection.java:1780)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:54)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:492)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.epcRunTask(AdaptiveExecutionStrategy.java:428)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:401)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:255)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:204)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:317)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:1009)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1240)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1194)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "PROJECT_NAME_VERSION_IDX"
  Detail: Key ("NAME", "VERSION")=(name, main) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2904)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2589)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:431)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:534)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:437)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:196)
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:157)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
	at org.datanucleus.store.rdbms.SQLController.doExecuteStatementUpdate(SQLController.java:465)
	at org.datanucleus.store.rdbms.SQLController.executeStatementUpdateDeferRowCountCheckForBatching(SQLController.java:415)
	at org.datanucleus.store.rdbms.request.InsertRequest.execute(InsertRequest.java:532)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.insertObjectInTable(RDBMSPersistenceHandler.java:235)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.insertObject(RDBMSPersistenceHandler.java:211)
	at org.datanucleus.state.StateManagerImpl.internalMakePersistent(StateManagerImpl.java:4610)
	at org.datanucleus.state.StateManagerImpl.makePersistent(StateManagerImpl.java:4587)
	at org.datanucleus.ExecutionContextImpl.persistObjectInternal(ExecutionContextImpl.java:2077)
	at org.datanucleus.ExecutionContext.persistObjectInternal(ExecutionContext.java:320)
	at org.datanucleus.ExecutionContextImpl.persistObjectWork(ExecutionContextImpl.java:1925)
	at org.datanucleus.ExecutionContextImpl.persistObject(ExecutionContextImpl.java:1786)
	at org.datanucleus.api.jdo.JDOPersistenceManager.jdoMakePersistent(JDOPersistenceManager.java:697)
	... 69 common frames omitted
```
> `name` was used to replace the actual project name.

The database logs mirror this accordingly:
```
2026-06-11 07:07:21.410 UTC [79861] ERROR:  duplicate key value violates unique constraint "PROJECT_NAME_VERSION_IDX"
2026-06-11 07:07:21.410 UTC [79861] DETAIL:  Key ("NAME", "VERSION")=(name, main) already exists.
2026-06-11 07:07:21.410 UTC [79861] STATEMENT:  INSERT INTO "PROJECT" ("AUTHORS","CLASSIFIER","COLLECTION_LOGIC","COLLECTION_TAG_ID","CPE","DESCRIPTION","DIRECT_DEPENDENCIES","EXTERNAL_REFERENCES","GROUP","INACTIVE_SINCE","IS_LATEST","LAST_BOM_IMPORTED","LAST_BOM_IMPORTED_FORMAT","LAST_RISKSCORE","LAST_VULNERABILITY_ANALYSIS","MANUFACTURER","NAME","PARENT_PROJECT_ID","PUBLISHER","PURL","SUPPLIER","SWIDTAGID","UUID","VERSION") VALUES ($1,$2,$3,$4,$5,$6,CAST($7 AS JSONB),$8,$9,$10,$11,$12,$13,$14,$15,$16,$17,$18,$19,$20,$21,$22,$23,$24)
	RETURNING "ID"
```

The `/v1/bom` API documentation states:
> Expects CycloneDX and a valid project UUID. If a UUID is not specified, then the `projectName` and `projectVersion` must be specified.
> Optionally, if `autoCreate` is specified and `true` and the project does not exist, the project will be created.

Now as to my understanding, uploading a BOM for an existing project by name+version should work the same as by uuid - just as it did before.
However, the current behavior seems to suggest that lookup does not seem to work quite as I imagined it in this case.
I also could not find any reference to behavior changes in this regard in the [list of breaking changes](https://dependencytrack.github.io/docs/next/reference/api/v5-breaking-changes/), so I assume this might be unintentional.
This being reported as an unhandled error seems to support that assumption.

**Note**: As outlined [here](https://dependencytrack.github.io/docs/next/tutorials/rehearsing-the-v4-migration/), the instance was migrated from a live V4 instance accepting potential inconsistencies, as the main purpose is load testing and scaling appropriately. Should this be a result thereof, feel free to close the issue.

### Steps to Reproduce

1. Migrate from v4 to v5
2. Set up BOM forwarding for parallel operation
3. Observe error for some requests

### Expected Behavior

BOM uploads that were successfully processed by V4 should also work on V5.

### Dependency-Track Version

5.x

### Browser

Google Chrome

### Checklist

- [x] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/main/CONTRIBUTING.md#filing-issues)
- [x] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this defect was already reported

--- a/apiserver/src/main/java/org/dependencytrack/resources/v1/BomResource.java
+++ b/apiserver/src/main/java/org/dependencytrack/resources/v1/BomResource.java
@@ -120,6 +120,7 @@ import static org.dependencytrack.model.ConfigPropertyConstants.BOM_VALIDATION_T
 import static org.dependencytrack.notification.api.NotificationFactory.createBomValidationFailedNotification;
 import static org.dependencytrack.notification.api.NotificationFactory.createProjectCreatedNotification;
 import static org.dependencytrack.persistence.jdbi.JdbiFactory.useJdbiTransaction;
+import static org.dependencytrack.util.PersistenceUtil.isUniqueConstraintViolation;
 
 /**
  * JAX-RS resources for processing bill-of-material (bom) documents.
@@ -189,9 +190,9 @@ public class BomResource extends AbstractApiResource {
             @QueryParam("version") String version
     ) {
         try (QueryManager qm = new QueryManager(getAlpineRequest())) {
-            String versionParameter =  Objects.toString(StringUtils.trimToNull(version), DEFAULT_EXPORT_VERSION);
+            String versionParameter = Objects.toString(StringUtils.trimToNull(version), DEFAULT_EXPORT_VERSION);
             Version cdxOutputVersion = Version.fromVersionString(versionParameter);
-            if(cdxOutputVersion == null) {
+            if (cdxOutputVersion == null) {
                 return Response.status(Response.Status.BAD_REQUEST).entity("Invalid BOM version specified.").build();
             }
 
@@ -397,7 +398,16 @@ public class BomResource extends AbstractApiResource {
             );
             try (final var qm = new QueryManager(getAlpineRequest())) {
                 projectInfo = qm.callInTransaction(() -> {
-                    Project project = qm.getProject(request.getProjectName(), request.getProjectVersion());
+                    final String trimmedProjectName = StringUtils.trimToNull(request.getProjectName());
+                    final String trimmedProjectVersion = StringUtils.trimToNull(request.getProjectVersion());
+
+                    // NB: Bypass portfolio ACL for this lookup since it would otherwise filter out existing
+                    // projects that are inaccessible. `autoCreate=true` would then lead to us trying to create
+                    // a project that already exists, triggering a unique constraint violation.
+                    // Access to the project (and potentially its parent) is asserted explicitly via requireAccess().
+                    Project project = ProjectAccess.unrestricted(
+                            () -> qm.getProject(trimmedProjectName, trimmedProjectVersion));
+
                     if (project == null && request.isAutoCreate()) {
                         if (hasPermission(Permissions.Constants.PORTFOLIO_MANAGEMENT) || hasPermission(Permissions.Constants.PORTFOLIO_MANAGEMENT_CREATE) || hasPermission(Permissions.Constants.PROJECT_CREATION_UPLOAD)) {
                             Project parent = null;
@@ -412,7 +422,7 @@ public class BomResource extends AbstractApiResource {
                                     );
                                     final String trimmedParentName = StringUtils.trimToNull(request.getParentName());
                                     final String trimmedParentVersion = StringUtils.trimToNull(request.getParentVersion());
-                                    parent = qm.getProject(trimmedParentName, trimmedParentVersion);
+                                    parent = ProjectAccess.unrestricted(() -> qm.getProject(trimmedParentName, trimmedParentVersion));
                                 }
 
                                 if (parent == null) {
@@ -423,16 +433,25 @@ public class BomResource extends AbstractApiResource {
                                 }
                                 requireAccess(qm, parent, "Access to the specified parent project is forbidden");
                             }
-                            final String trimmedProjectName = StringUtils.trimToNull(request.getProjectName());
                             if (request.isLatest()) {
                                 final Project oldLatest = ProjectAccess.unrestricted(() -> qm.getLatestProjectVersion(trimmedProjectName));
                                 if (oldLatest != null) {
                                     requireAccess(qm, oldLatest, "Access to the previous latest project version is forbidden");
                                 }
                             }
-                            project = qm.createProject(trimmedProjectName, null,
-                                    StringUtils.trimToNull(request.getProjectVersion()), request.getProjectTags(), parent,
-                                    null, null, request.isLatest(), true);
+                            try {
+                                project = qm.createProject(trimmedProjectName, null,
+                                        trimmedProjectVersion, request.getProjectTags(), parent,
+                                        null, null, request.isLatest(), true);
+                            } catch (RuntimeException e) {
+                                if (isUniqueConstraintViolation(e)) {
+                                    throw new WebApplicationException(Response
+                                            .status(Response.Status.CONFLICT)
+                                            .entity("A project with the specified name and version already exists.")
+                                            .build());
+                                }
+                                throw e;
+                            }
                             Principal principal = getPrincipal();
                             qm.updateNewProjectACL(project, principal);
                             new JdoNotificationEmitter(qm).emit(
@@ -574,7 +593,14 @@ public class BomResource extends AbstractApiResource {
                 projectInfo = qm.callInTransaction(() -> {
                     final String trimmedProjectName = StringUtils.trimToNull(projectName);
                     final String trimmedProjectVersion = StringUtils.trimToNull(projectVersion);
-                    Project project = qm.getProject(trimmedProjectName, trimmedProjectVersion);
+
+                    // NB: Bypass portfolio ACL for this lookup since it would otherwise filter out existing
+                    // projects that are inaccessible. `autoCreate=true` would then lead to us trying to create
+                    // a project that already exists, triggering a unique constraint violation.
+                    // Access to the project (and potentially its parent) is asserted explicitly via requireAccess().
+                    Project project = ProjectAccess.unrestricted(
+                            () -> qm.getProject(trimmedProjectName, trimmedProjectVersion));
+
                     if (project == null && autoCreate) {
                         if (hasPermission(Permissions.Constants.PORTFOLIO_MANAGEMENT) || hasPermission(Permissions.Constants.PORTFOLIO_MANAGEMENT_CREATE) || hasPermission(Permissions.Constants.PROJECT_CREATION_UPLOAD)) {
                             Project parent = null;
@@ -584,7 +610,7 @@ public class BomResource extends AbstractApiResource {
                                 } else {
                                     final String trimmedParentName = StringUtils.trimToNull(parentName);
                                     final String trimmedParentVersion = StringUtils.trimToNull(parentVersion);
-                                    parent = qm.getProject(trimmedParentName, trimmedParentVersion);
+                                    parent = ProjectAccess.unrestricted(() -> qm.getProject(trimmedParentName, trimmedParentVersion));
                                 }
 
                                 if (parent == null) {
@@ -601,7 +627,17 @@ public class BomResource extends AbstractApiResource {
                                     requireAccess(qm, oldLatest, "Access to the previous latest project version is forbidden");
                                 }
                             }
-                            project = qm.createProject(trimmedProjectName, null, trimmedProjectVersion, requestTags, parent, null, null, isLatest, true);
+                            try {
+                                project = qm.createProject(trimmedProjectName, null, trimmedProjectVersion, requestTags, parent, null, null, isLatest, true);
+                            } catch (RuntimeException e) {
+                                if (isUniqueConstraintViolation(e)) {
+                                    throw new WebApplicationException(Response
+                                            .status(Response.Status.CONFLICT)
+                                            .entity("A project with the specified name and version already exists.")
+                                            .build());
+                                }
+                                throw e;
+                            }
                             Principal principal = getPrincipal();
                             qm.updateNewProjectACL(project, principal);
                             new JdoNotificationEmitter(qm).emit(

--- a/apiserver/src/test/java/org/dependencytrack/resources/v1/BomResourceTest.java
+++ b/apiserver/src/test/java/org/dependencytrack/resources/v1/BomResourceTest.java
@@ -2055,6 +2055,173 @@ class BomResourceTest extends ResourceTest {
     }
 
     @Test
+    void shouldReturnForbiddenWhenAutoCreateButProjectExistsInaccessible() throws Exception {
+        initializeWithPermissions(Permissions.BOM_UPLOAD, Permissions.PROJECT_CREATION_UPLOAD);
+        enablePortfolioAccessControl();
+
+        final var existing = new Project();
+        existing.setName("acme-app");
+        existing.setVersion("1.0.0");
+        qm.persist(existing);
+
+        final String bomString = Base64.getEncoder().encodeToString(resourceToByteArray("/unit/bom-1.xml"));
+        final Response response = jersey.target(V1_BOM).request()
+                .header(X_API_KEY, apiKey)
+                .put(Entity.json(/* language=JSON */ """
+                        {
+                          "projectName": "acme-app",
+                          "projectVersion": "1.0.0",
+                          "autoCreate": true,
+                          "bom": "%s"
+                        }
+                        """.formatted(bomString)));
+        assertThat(response.getStatus()).isEqualTo(403);
+
+        final Project found = qm.getProject("acme-app", "1.0.0");
+        assertThat(found).isNotNull();
+        assertThat(found.getUuid()).isEqualTo(existing.getUuid());
+    }
+
+    @Test
+    void shouldReturnForbiddenWhenMultipartAutoCreateButProjectExistsInaccessible() throws Exception {
+        initializeWithPermissions(Permissions.BOM_UPLOAD, Permissions.PROJECT_CREATION_UPLOAD);
+        enablePortfolioAccessControl();
+
+        final var existing = new Project();
+        existing.setName("Acme Example");
+        existing.setVersion("1.0");
+        qm.persist(existing);
+
+        final var multiPart = new FormDataMultiPart()
+                .field("bom", resourceToString("/unit/bom-1.xml", StandardCharsets.UTF_8), MediaType.APPLICATION_XML_TYPE)
+                .field("projectName", "Acme Example")
+                .field("projectVersion", "1.0")
+                .field("autoCreate", "true");
+
+        final var client = ClientBuilder.newClient(
+                new ClientConfig()
+                        .register(MultiPartFeature.class)
+                        .connectorProvider(new HttpUrlConnectorProvider()));
+
+        final Response response = client
+                .target(jersey.target(V1_BOM).getUri())
+                .request()
+                .header(X_API_KEY, apiKey)
+                .post(Entity.entity(multiPart, multiPart.getMediaType()));
+        assertThat(response.getStatus()).isEqualTo(403);
+
+        final Project found = qm.getProject("Acme Example", "1.0");
+        assertThat(found).isNotNull();
+        assertThat(found.getUuid()).isEqualTo(existing.getUuid());
+    }
+
+    @Test
+    void shouldMatchExistingProjectWhenNameAndVersionContainSurroundingWhitespace() throws Exception {
+        initializeWithPermissions(Permissions.BOM_UPLOAD, Permissions.PROJECT_CREATION_UPLOAD);
+
+        final var existing = new Project();
+        existing.setName("acme-app");
+        existing.setVersion("1.0.0");
+        qm.persist(existing);
+
+        final String bomString = Base64.getEncoder().encodeToString(resourceToByteArray("/unit/bom-1.xml"));
+        final Response response = jersey
+                .target(V1_BOM)
+                .request()
+                .header(X_API_KEY, apiKey)
+                .put(Entity.json(/* language=JSON */ """
+                        {
+                          "projectName": "  acme-app  ",
+                          "projectVersion": "  1.0.0  ",
+                          "autoCreate": true,
+                          "bom": "%s"
+                        }
+                        """.formatted(bomString)));
+        assertThat(response.getStatus()).isEqualTo(200);
+
+        final Project found = qm.getProject("acme-app", "1.0.0");
+        assertThat(found).isNotNull();
+        assertThat(found.getUuid()).isEqualTo(existing.getUuid());
+    }
+
+    @Test
+    void shouldReturnForbiddenWhenParentProjectByNameIsInaccessible() throws Exception {
+        initializeWithPermissions(Permissions.BOM_UPLOAD, Permissions.PROJECT_CREATION_UPLOAD);
+        enablePortfolioAccessControl();
+
+        final var parent = new Project();
+        parent.setName("Acme Parent");
+        parent.setVersion("1.0");
+        qm.persist(parent);
+
+        final String bomString = Base64.getEncoder().encodeToString(resourceToByteArray("/unit/bom-1.xml"));
+        final Response response = jersey
+                .target(V1_BOM)
+                .request()
+                .header(X_API_KEY, apiKey)
+                .put(Entity.json(/* language=JSON */ """
+                        {
+                          "projectName": "Acme Example",
+                          "projectVersion": "1.0",
+                          "parentName": "Acme Parent",
+                          "parentVersion": "1.0",
+                          "autoCreate": true,
+                          "bom": "%s"
+                        }
+                        """.formatted(bomString)));
+        assertThat(response.getStatus()).isEqualTo(403);
+        assertThatJson(getPlainTextBody(response)).isEqualTo(/* language=JSON */ """
+                {
+                  "status": 403,
+                  "title": "Project access denied",
+                  "detail": "Access to the specified parent project is forbidden"
+                }
+                """);
+
+        assertThat(qm.getProject("Acme Example", "1.0")).isNull();
+    }
+
+    @Test
+    void shouldReturnForbiddenWhenMultipartParentProjectByNameIsInaccessible() throws Exception {
+        initializeWithPermissions(Permissions.BOM_UPLOAD, Permissions.PROJECT_CREATION_UPLOAD);
+        enablePortfolioAccessControl();
+
+        final var parent = new Project();
+        parent.setName("Acme Parent");
+        parent.setVersion("1.0");
+        qm.persist(parent);
+
+        final var multiPart = new FormDataMultiPart()
+                .field("bom", resourceToString("/unit/bom-1.xml", StandardCharsets.UTF_8), MediaType.APPLICATION_XML_TYPE)
+                .field("projectName", "Acme Example")
+                .field("projectVersion", "1.0")
+                .field("parentName", "Acme Parent")
+                .field("parentVersion", "1.0")
+                .field("autoCreate", "true");
+
+        final var client = ClientBuilder.newClient(
+                new ClientConfig()
+                        .register(MultiPartFeature.class)
+                        .connectorProvider(new HttpUrlConnectorProvider()));
+
+        final Response response = client
+                .target(jersey.target(V1_BOM).getUri())
+                .request()
+                .header(X_API_KEY, apiKey)
+                .post(Entity.entity(multiPart, multiPart.getMediaType()));
+        assertThat(response.getStatus()).isEqualTo(403);
+        assertThatJson(getPlainTextBody(response)).isEqualTo(/* language=JSON */ """
+                {
+                  "status": 403,
+                  "title": "Project access denied",
+                  "detail": "Access to the specified parent project is forbidden"
+                }
+                """);
+
+        assertThat(qm.getProject("Acme Example", "1.0")).isNull();
+    }
+
+    @Test
     void shouldRejectBomUploadForCollectionProject() throws Exception {
         initializeWithPermissions(Permissions.BOM_UPLOAD);
         final var project = new Project();

